src/serialization/compact/DefaultCompactReader.ts

```diff
@@ -0,0 +1,762 @@
+import {CompactGenericRecordImpl} from '../generic_record';
+import {Field} from '../generic_record/Fields';
+import {
+    BYTE_OFFSET_READER,
+    BYTE_OFFSET_READER_RANGE,
+    INT_OFFSET_READER, NULL_OFFSET,
+    SHORT_OFFSET_READER,
+    SHORT_OFFSET_READER_RANGE
+} from './OffsetConstants';
+import {BitsUtil} from '../../util/BitsUtil';
+import {OffsetReader} from './OffsetReader';
+import {FieldDescriptor} from '../generic_record/FieldDescriptor';
+import {CompactExceptions} from './CompactUtil';
+import {IOUtil} from '../../util/IOUtil';
+import {FieldOperations} from '../generic_record/FieldOperations';
+
+/**
+ * Represents unserialized form of a compact object. Users do not receive this object.
+ * Instead, they receive {@link CompactGenericRecordImpl}
+ *
+ * @internal
+ */
+export class DefaultCompactReader implements CompactReader {
+    private readonly offsetReader: OffsetReader;
+    private readonly variableOffsetsPosition: number;
+    private readonly dataStartPosition: number;
+
+    constructor(
+        private readonly serializer: CompactStreamSerializer,
+        private readonly input: ObjectDataInput,
+        private readonly schema: Schema,
+    ) {
+        const numberOfVariableLengthFields = schema.numberVarSizeFields;
+        let finalPosition: number;
+        if (numberOfVariableLengthFields !== 0) {
+            const dataLength = input.readInt();
+            this.dataStartPosition = this.input.position();
+            this.variableOffsetsPosition = this.dataStartPosition + dataLength;
+            if (dataLength < BYTE_OFFSET_READER_RANGE) {
+                this.offsetReader = BYTE_OFFSET_READER;
+                finalPosition = this.variableOffsetsPosition + numberOfVariableLengthFields;
+            } else if (dataLength < SHORT_OFFSET_READER_RANGE) {
+                this.offsetReader = SHORT_OFFSET_READER;
+                finalPosition = this.variableOffsetsPosition + numberOfVariableLengthFields * BitsUtil.SHORT_SIZE_IN_BYTES;
+            } else {
+                this.offsetReader = INT_OFFSET_READER;
+                finalPosition = this.variableOffsetsPosition + numberOfVariableLengthFields * BitsUtil.INT_SIZE_IN_BYTES;
+            }
+        } else {
+            this.offsetReader = INT_OFFSET_READER;
+            this.variableOffsetsPosition = 0;
+            this.dataStartPosition = input.position();
+            finalPosition = this.dataStartPosition + schema.fixedSizeFieldsLength;
+        }
+        //set the position to final so that the next one to read something from `input` can start from
+        //correct position
+        this.input.position(finalPosition);
+    }
+
+    getFieldKind(fieldName: string): FieldKind {
+        const fd = this.schema.fieldDefinitionMap.get(fieldName);
+        if (fd === undefined) {
+            return FieldKind.NOT_AVAILABLE;
+        }
+        return fd.kind;
+    }
+
+    readBoolean(fieldName: string): boolean {
```

Comment:
> we don't have the mechanism you have were you can use one method or the other.

What do you mean? I think with getFieldKind you can do that.

```js
let bool;

if (reader.getFieldKind('fieldName') === FieldKind.BOOLEAN) {
    bool = reader.readBoolean('fieldName');
} else {
   bool = reader.readNullableBoolean('fieldName');
}
```

Do you mean this one?

Reply:
yes this one  - in .NET (and I believe, Java) if you use `readBoolean` to read a nullable boolean, and vice-versa, it throws?

Reply:
update: Java has the same implicit conversion mechanism because Java compact reader derives from Java generic record implementation, which has that implicit conversion mechanism built-in. this should be documented in some specs really. anyways it means it's OK here, and the .NET client will have to be updated.